enhancements/dns/forwarding.md

```diff
@@ -0,0 +1,246 @@
+---
+title: configurable-dns-forwarding
+authors:
+  - "@dhansen"
+reviewers:
+  - "@ironcladlou"
+  - "@Miciah"
+  - "@frobware"
+approvers:
+  - "@knobunc"
+  - "@ironcladlou"
+creation-date: 2019-09-27
+last-updated: 2019-09-27
+status: implementable
+see-also: 
+replaces:
+superseded-by:
+---
+
+# Configurable DNS Forwarding
+
+This proposal allows cluster operators the ability to configure DNS name query forwarding.
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Open Questions [optional]
+
+1. Is `operator.openshift.io` the best API group for the proposed `Upstream` type?
+2. How are API changes handled for cluster downgrades?
+3. Should IPv6 references for `Nameservers` be removed until OpenShift adds IPv6 support?
+
+## Summary
+
+Currently, the [cluster-dns-operator](https://github.com/openshift/cluster-dns-operator) manages DNS forwarding with a
+static ConfigMap:
+
+```bash
+$ oc get configmap/dns-default -n openshift-dns -o yaml
+apiVersion: v1
+data:
+  Corefile: |
+    .:5353 {
+        errors
+        health
+        kubernetes cluster.local in-addr.arpa ip6.arpa {
+            pods insecure
+            upstream
+            fallthrough in-addr.arpa ip6.arpa
+        }
+        prometheus :9153
+        forward . /etc/resolv.conf
+        cache 30
+        reload
+    }
+kind: ConfigMap
+metadata:
+  labels:
+    dns.operator.openshift.io/owning-dns: default
+  name: dns-default
+  namespace: openshift-dns
+  <SNIP>
+```
+
+This proposal will generate this ConfigMap with a `forward` section based on user-provided values of type `Upstream`.
+
+## Motivation
+
+DNS name resolution for pods is currently provided by an instance of [CoreDNS](https://coredns.io) that runs on each
+node in the cluster. If a DNS query is for a hostname within the cluster domain suffix (i.e. `cluster.local`), CoreDNS
+resolves the name. Otherwise, CoreDNS proxies DNS messages to an upstream resolver identified by `/etc/resolv.conf` on
+the corresponding node. Although this provides a consistent and reliable approach for name resolution, it restricts how
+cluster operators manage DNS name queries.
+
+### Goals
+
+1. A well defined API for DNS forwarding configuration that is implementation agnostic.
+2. A minimal API surface that can be expanded to support future DNS forwarding use cases.
+
+### Non-Goals
+
+1. Support every possible DNS forwarding use case.
+2. Configure or manage external DNS providers.
+
+## Proposal
+
+The proposed `Upstream` type defines a schema for associating one or more `Nameservers` to a DNS subdomain identified by
+`Domain`. `Nameservers` are responsible for resolving name queries for `Domain`:
+```go
+type Upstream struct {
+	Domain string `json:"domain"`
```

Comment:
I thought for the first rev of this we'd just forward all requests to the nameserver list rather than supporting domain-based routing. Thoughts?

Reply:
I think domain-based routing is fundamental to the enhancement.  If the need were only to configure upstreams, the solution would be to configure them in `resolv.conf` (or in DHCP whence the entries in `resolv.conf` come).

Reply:
>If the need were only to configure upstreams, the solution would be to configure them in resolv.conf

How can you configure resolv.conf on nodes in OpenShift IPI? I thought the fact you can't is one of the motivating factors.

>(or in DHCP whence the entries in resolv.conf come).

Again, how? Customers _were_ doing this and can't now, AFAIK.

Looking back through other non-public use case materials, I think this requirement may have shifted. Originally it was simply a `[]string` of name servers to which `*` would be forwarded., but I do see some chatter about forwarding a _single_ domain to a list, which of course might as well me represented as a map to handle the general case. Whether the single domain list would be satisfied by a wildcard, I don't know.

Anyway, no objection to a mapping, I just wanted to make sure we weren't making it more complex than necessary based on use cases we actually know of.

Reply:
> How can you configure resolv.conf on nodes in OpenShift IPI? I thought the fact you can't is one of the motivating factors.
> 
> > (or in DHCP whence the entries in resolv.conf come).
> 
> Again, how? Customers _were_ doing this and can't now, AFAIK.

On environments without DHCP, use machine-config-operator to create `/etc/resolv.conf`.  On environments with DHCP, configure DHCP correctly.  We shouldn't need to expand our API to provide a workaround for broken DHCP.